### ISSUE TITLE
Question(weight quantization): Why not write out fp8 after performing weight update?

### ISSUE THREAD

**Issue by jon-chuang** [NVIDIA/TransformerEngine#424]
In the GTC talk, it is mentioned that the fp8 quantized weights are precomputed (cached) only for microbatch case such as pipelining.

However, why not just write out fp8 quantized weights during the weight update stage? 

By contrast, perfoming a cast step requires a high-bandwidth read and write step prior to the matmul, unless the cast is recomputed in every tile.

This can also potentially save bandwidth even while updating the weights, in addition to during the forward pass. 

Furthermore, storing the weights and optimizer states in FP8 will save memory (including `m, v` - 1st and 2nd moments for Adam)

@ptrendx

**Comment by timmoon10** [comment]
Just to clarify, you are thinking of the optimizer step? The optimizer still needs to operate on FP32 params since even FP16/BF16 wil degrade learning quality. A fused kernel that combines the FP32 optimizer step and an FP8 cast would definitely be nice since it avoids an extra 4B/param read, but handling the FP8 scaling factors makes it a bit tricky.

**Comment by jon-chuang** [comment]
Right, it does seem like a high engineering burden to implement it in CUDA. I am thinking that one could implement it in something like Triton. 

However, there would also be a need to define a common API for TE in terms of this optimizer. It seems wise to at least run a benchmark on a single training step to see if this can improve the training step time and by how much.

**Comment by jon-chuang** [comment]
Btw, may I ask if you have profiled the transformer layer? Is there any public profiling available at various parameter choices (the proportional time take for layer norm, MLP, dot product attention, projection steps). If not I'll try to run Nsight systems at various problem sizes to try to get an idea (I guess that at large seq_len, attention would dominate, but its a good question what the bottleneck is at small or moderate seq len).

**Comment by nouiz** [comment]
The transformer layer have been well studied. If you looks at commits, you will see many changes related to the Multi-Head Attention.

**Comment by jon-chuang** [comment]
After some exploration, I found that the use of FP8 beyond matmul has been studied in Microsoft's AMP project ([repo](https://github.com/Azure/MS-AMP)). It is used for:
1. Optimizer states (Adam, AdamW)
2. Weight updates

These are the results found by them:
> Specifically, we are able to maintain accuracy by representing the first-order optimizer state in FP8 and the second-order state in FP16.

Here is their [implementation](https://github.com/Azure/MS-AMP/blob/aed29d6533e8ff86686fdb6fafa6d0b720e9e5f6/msamp/optim/adamw.cu) of AdamW in CUDA.

---

Would be great to get your inputs @tocean @wkcn. 

Questions: 
1. The repo only mention training accuracy and memory savings. However, the kernels were not very optimized and majority is implemented in Torch. I guess that performance is still unexplored.
2. For weight update:
    - is the weight update applied while the backward pass is running (on-the-fly)? Or is it applied after the entire backward pass is complete?
3. Is there a way to maintain more accurate amax by estimating `amax_weights_t = amax_weights_t-1 + amax_grad_t`?

**Comment by ptrendx** [comment]
Hi @jon-chuang - having the optimizer write out FP8 is a good idea and we are looking into this (having the state itself in FP8 is a different story - we prefer to be conservative and move slower when it comes to changes that affect the numerical precision of the computation). 
That said, the first step to achieve this (at least for pyTorch) is to have an actual FP8 weights in the model (as opposed to storing the weights in a regular FP32 or 16-bit tensor). This would reduce the memory usage and allow the optimizer to interact with those values directly. We are currently working on supporting this as well, so stay tuned :-).

**Comment by jon-chuang** [comment]
Awesome @ptrendx. In the meantime, I'll see if I can get some experiments going in Triton for storing the weights and optimizer states as FP8, in addition to computations (I might be focusing on JAX side so it may be somewhat orthogonal to Pytorch efforts!).

**Comment by sbhavani** [comment]
Thanks for raising this @jon-chuang. Since this issue was opened, TE has added support for quantized primary weights via `quantized_model_init`, low-precision/FP8 optimizer states in `FusedAdam` (#1078), FP8 primary weights in FSDP training (#1630), and broader quantized-weight + `FusedAdam` integration (#2698, #2753, #2797).

For delayed-scaling `Float8Tensor` parameters, `FusedAdam` now covers the main optimization discussed here by updating the FP32 master weights and writing the FP8 parameter buffer/metadata during the optimizer step. Other quantized formats currently update FP32 master weights and then requantize/write back afterward, so further fusion there could still be future performance work.

Closing this as substantially addressed by the newer quantized primary weight and fused optimizer support.
